src/axolotl/train.py

```diff
@@ -187,32 +188,49 @@ def terminate_handler(_, __, model_weakref):
     else:
         trainer.train(resume_from_checkpoint=resume_from_checkpoint)
 
-    post_train_hooks(cfg, trainer)
 
+def save_trained_model(
+    cfg: DictDefault,
+    trainer: Any,
+    model: PreTrainedModel,
+    safe_serialization: bool,
+):
+    """
+    Save the trained model according to configuration and training setup.
+
+    Args:
+        cfg: The configuration dictionary
+        trainer: The trainer object
+        model: The trained model to save
+        safe_serialization: Whether to use safe serialization
+    """
     LOG.info(f"Training Completed!!! Saving pre-trained model to {cfg.output_dir}")
 
-    # post training
+    # Post training module hooks
     for name, module in model.named_modules():
         if hasattr(module, "_post_training"):
             module._post_training(model, name)  # pylint: disable=protected-access
 
+    # Handle FSDP state dict type
     state_dict_type = "FULL_STATE_DICT"
     if trainer.is_fsdp_enabled:
         if cfg.fsdp_final_state_dict_type:
             state_dict_type = cfg.fsdp_final_state_dict_type
         trainer.accelerator.state.fsdp_plugin.set_state_dict_type(state_dict_type)
         LOG.info(f"Set FSDP state dict type to {state_dict_type} for saving.")
 
+    # Handle ReLoRA early return case
     if cfg.relora_steps:
         if cfg.adapter == "lora" and not (cfg.load_in_4bit or cfg.load_in_8bit):
             model = model.merge_and_unload()
         else:
             # final model weights have already been saved by `ReLoRACallback.on_train_end`
-            return model, tokenizer
+            return
 
-    # TODO do we need this fix? https://huggingface.co/docs/accelerate/usage_guides/fsdp#saving-and-loading
-    # only save on rank 0, otherwise it corrupts output on multi-GPU when multiple processes attempt to write the same file
     if cfg.fsdp:
+        # TODO: do we need this fix? https://huggingface.co/docs/accelerate/usage_guides/fsdp#saving-and-loading
+        # only save on rank 0, otherwise it corrupts output on multi-GPU when multiple
+        # processes attempt to write the same file
```

Comment:
Re: this TODO, do we need to call their `merge_fsdp_weights` instead?

Reply:
No idea, just copying this TODO from the existing code. We could make a ticket to look into this.